piwik-sdk/src/main/java/org/piwik/sdk/TrackHelper.java

```diff
@@ -0,0 +1,719 @@
+    }
+
+    private TrackHelper(@Nullable TrackMe baseTrackMe) {
+        if (baseTrackMe == null) baseTrackMe = new TrackMe();
+        mBaseTrackMe = baseTrackMe;
+    }
+
+    public static TrackHelper track() {
+        return new TrackHelper();
+    }
+
+    public static TrackHelper track(@NonNull TrackMe base) {
+        return new TrackHelper(base);
+    }
+
+    static abstract class BaseEvent {
+
+        private final TrackHelper mBaseBuilder;
+
+        BaseEvent(TrackHelper baseBuilder) {
+            mBaseBuilder = baseBuilder;
+        }
+
+        TrackMe getBaseTrackMe() {
+            return mBaseBuilder.mBaseTrackMe;
+        }
+
+        @Nullable
+        public abstract TrackMe build();
+
+        public void with(@NonNull PiwikApplication piwikApplication) {
+            with(piwikApplication.getTracker());
+        }
+
+        public void with(@NonNull Tracker tracker) {
+            TrackMe trackMe = build();
+            if (trackMe != null) tracker.track(trackMe);
+        }
+    }
+
+    /**
+     * To track a screenview.
+     *
+     * @param path Example: "/user/settings/billing"
+     * @return an object that allows addition of further details.
+     */
+    public Screen screen(String path) {
+        return new Screen(this, path);
+    }
+
+    /**
+     * Calls {@link #screen(String)} for an activity.
+     * Uses the activity-stack as path and activity title as names.
+     *
+     * @param activity the activity to track
+     */
+    public Screen screen(Activity activity) {
+        String breadcrumbs = ActivityHelper.getBreadcrumbs(activity);
+        return new Screen(this, ActivityHelper.breadcrumbsToPath(breadcrumbs)).title(breadcrumbs);
+    }
+
+    public static class Screen extends BaseEvent {
+        private final String mPath;
+        private final CustomVariables mCustomVariables = new CustomVariables();
+        private String mTitle;
+
+        Screen(TrackHelper baseBuilder, String path) {
+            super(baseBuilder);
+            mPath = path;
+        }
+
+        /**
+         * The title of the action being tracked. It is possible to use slashes / to set one or several categories for this action.
+         *
+         * @param title Example: Help / Feedback will create the Action Feedback in the category Help.
+         * @return this object to allow chaining calls
+         */
+        public Screen title(String title) {
+            mTitle = title;
+            return this;
+        }
+
+        /**
+         * Just like {@link Tracker#setVisitCustomVariable(int, String, String)} but only valid per screen.
+         * Only takes effect when setting prior to tracking the screen view.
+         */
+        public Screen variable(int index, String name, String value) {
+            mCustomVariables.put(index, name, value);
+            return this;
+        }
+
+        @Nullable
+        @Override
+        public TrackMe build() {
+            if (mPath == null) return null;
```

Comment:
In this case `path` is the **required** variable for tracking a screenview, which is why we want to "abort" if it is null.

At somepoint we have or should give an oportunity for sanity checks on the values, although we currently don't have many of those, there needs to be a place when to do so.
As we are moving the TrackMe and Tracker class to pure information carriers, the "Builder" has to do this job, either when calling the parameter functions or when calling build(), in this case i've chosen the later one, because this gives us the option to do sanity checks against parameters form different methods the dev might have called.

The idea is that build() returns a null value when it is not a valid (use senseless) request to submit for transmission. Returning `null` means that some values/*whatever* is fishy and the TrackMe that the developer tried to generate is likely invalid.

I could see two reasons for never returning a `null`-value here but both would be contrary to what we want. So for one if we return the base `TrackMe` we could further chain calls, but for readabilities sake, we want to encourage one callchain per tracking request.

It should also be noted that the "mBaseTrack" is not modified by a request. Specifically the "sub builders" such as `Screen` in this instance, don't modify the base `TrackMe` object. The idea is to allow the developer to keep a base TrackMe object as a global field with certain parameters (e.g. some plugin stuff), that he then can always uses in the builders with TrackHelper.track(baseTrackMe).

Also see extension of the `TrackMe` constructor allowing us to use `new TrackMe(oldTrackMe)`.

The other reason we might want a non `null` value here is to avoid NPE issues, because the Tracker would crash if we do `track(null)`, but either the developer does a null check or he uses the convinience method `TrackHelper......with(tracker)` which internally calls `build()` and simply does nothing if it is `null`.

Need to add some documentation on this... will do that before this would be merged, but I'm happy to get some review on this first.